### PR TITLE
AdSense: only add AMP auto-ads via main loop `the_content` filter.

### PR DESCRIPTION
**Pull request by adamsilverstein**
## Summary

Addresses issue https://github.com/google/site-kit-wp/issues/2111

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Comment by adamsilverstein** [comment]
Note: ideally we should have tests that verify the correct tag is output in each of the three AMP modes, for the latest of each major version of the AMP plugin (1.x.x and 2.x.x).

**Comment by ryanwelcher** [comment]
@googlebot I consent.

File: tests/e2e/matchers/to-have-amp-auto-ads-tag.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+/**
```
Comment (aaemnnosttv) [inline tests/e2e/matchers/to-have-amp-auto-ads-tag.js:1]: Please add a file header 🙂

File: tests/e2e/plugins/apply-content-fliters.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ * Author URI:  https://opensource.google.com
+ *
+ * @package   Google\Site_Kit
+ * @copyright 2020 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+
+// Applying the filters outside of the loop breaks the AdSense snippet with AMP enabled.
```
Comment (aaemnnosttv) [inline tests/e2e/plugins/apply-content-fliters.php:15]: This comment is no longer accurate as of this PR, correct?
I think it would be good to reword this to be specific to this being a previous bug.
It would also be good to include a link to the issue here as well for additional context.

File: tests/e2e/specs/modules/adsense/setup-new-user.test.js [hunk 3]
```diff
@@ -345,4 +346,59 @@ describe( 'setting up the AdSense module', () => {
...
+				} );
+			};
+		} );
+
+		afterAll( async () => {
+			await deactivatePlugin( 'amp' );
+			await deactivatePlugin( 'e2e-tests-apply-content-filters' );
+		} );
+
+		it( 'it output in primary mode', async () => {
```
Comment (aaemnnosttv) [inline tests/e2e/specs/modules/adsense/setup-new-user.test.js:393]:
```suggestion
		it( 'is output in primary mode', async () => {
```

File: tests/e2e/specs/modules/adsense/setup-new-user.test.js [hunk 4]
```diff
@@ -345,4 +346,59 @@ describe( 'setting up the AdSense module', () => {
...
+			await setAMPMode( 'primary' );
+			await proceedToAdsenseSetup();
+			await expect( '/hello-world' ).toHaveAMPAutoAdsTag();
+		} );
+		it( 'it output in secondary mode', async () => {
```
Comment (aaemnnosttv) [inline tests/e2e/specs/modules/adsense/setup-new-user.test.js:398]:
```suggestion
		it( 'is output in secondary mode', async () => {
```

File: tests/e2e/specs/modules/adsense/setup-new-user.test.js [hunk 5]
```diff
@@ -345,4 +346,59 @@ describe( 'setting up the AdSense module', () => {
...
+		afterAll( async () => {
+			await deactivatePlugin( 'amp' );
+			await deactivatePlugin( 'e2e-tests-apply-content-filters' );
+		} );
+
+		it( 'it output in primary mode', async () => {
+			await setAMPMode( 'primary' );
+			await proceedToAdsenseSetup();
+			await expect( '/hello-world' ).toHaveAMPAutoAdsTag();
+		} );
```
Comment (aaemnnosttv) [inline tests/e2e/specs/modules/adsense/setup-new-user.test.js:397]:
```suggestion
		} );

```

File: tests/e2e/specs/modules/adsense/setup-new-user.test.js [hunk 6]
```diff
@@ -345,4 +346,59 @@ describe( 'setting up the AdSense module', () => {
...
 	} );
+
+	describe( 'AMP <amp-auto-ads> tag', () => {
+		beforeAll( async () => {
+			await activatePlugin( 'amp' );
+			await activatePlugin( 'e2e-tests-apply-content-filters' );
+		} );
+
+		beforeEach( () => {
+			datapointHandlers.accounts = ( request ) => {
```
Comment (aaemnnosttv) [inline tests/e2e/specs/modules/adsense/setup-new-user.test.js:357]: [Per @felixarntz 's comment](https://github.com/google/site-kit-wp/issues/2111#issuecomment-707305831) it is a bit inefficient to go through the AdSense setup for each AMP mode just to check the tag output.

How about we do the following:
* add a `setupAdSense` utility, similar to `setupAnalytics` in `tests/e2e/utils/setup-analytics.js` since these tests aren't concerned with the module setup at all
* Move this test to a new `amp-auto-ads-tag.test.js` file in the same directory
* Use the new utility to setup AdSense in the `beforeAll` instead of going through setup for each test

File: tests/phpunit/integration/Modules/AdSenseTest.php [hunk 7]
```diff
@@ -187,6 +187,7 @@ public function test_block_on_consent_amp( $enabled ) {
 
 	/**
 	 * @dataProvider block_on_consent_provider
+	 * @group ryan
```
Comment (aaemnnosttv) [inline tests/phpunit/integration/Modules/AdSenseTest.php:190]:
```suggestion
```

File: tests/phpunit/integration/Modules/AdSenseTest.php [hunk 8]
```diff
@@ -204,6 +205,10 @@ public function test_block_on_consent_amp_content( $enabled ) {
 			add_filter( 'googlesitekit_adsense_tag_amp_block_on_consent', '__return_true' );
 		}
 
+		// We need to fake the global to allow the hook to add the tag.
+		global $wp_query;
+		$wp_query->in_the_loop = true;
```
Comment (aaemnnosttv) [inline tests/phpunit/integration/Modules/AdSenseTest.php:210]: While this is necessary to make this test pass, let's add another test for this specifically to test that the tag is also not added when not in the loop.

File: bin/local-env/install-wordpress.sh [hunk 9]
```diff
@@ -113,6 +113,7 @@ else
 	wp plugin update amp --minor --quiet
 fi
 
+
```
Comment (aaemnnosttv) [inline bin/local-env/install-wordpress.sh:116]:
```suggestion
```

File: includes/Modules/AdSense.php [hunk 10]
```diff
@@ -132,6 +132,10 @@ function() use ( $client_id ) {
 					add_filter( // For AMP Reader, and AMP Native and Transitional (as fallback).
 						'the_content',
 						function( $content ) use ( $client_id ) {
+							// Only run for the primary application the `the_content` filter.
```
Comment (aaemnnosttv) [inline includes/Modules/AdSense.php:135]:
```suggestion
							// Only run for the primary application of the `the_content` filter.
```

**Review by aaemnnosttv — CHANGES_REQUESTED** [review]
Thanks @ryanwelcher – this is looking pretty good, just a few changes to make and minor things to clean up. The main thing being a change to the way the E2E tests run to be a bit more efficient.

File: tests/phpunit/integration/Modules/AdSenseTest.php (hunk 7 above)
Comment (ryanwelcher) [reply tests/phpunit/integration/Modules/AdSenseTest.php:190]: 🤦

File: tests/e2e/matchers/to-have-amp-auto-ads-tag.js [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+ * @since n.e.x.t
+ *
+ * @param {string} path The URL path of the current site to check.
+ * @return {Object} Matcher result.
+ */
+export async function toHaveAMPAutoAdsTag( path ) {
+	const result = {};
+
+	const page = await browser.newPage();
+	await page.goto( createURL( path, 'amp' ) );
```
Comment (aaemnnosttv) [inline tests/e2e/matchers/to-have-amp-auto-ads-tag.js:36]: Since the `amp` query param only applies to secondary AMP modes (traditional at least), let's pass that as part of the `path` when calling rather than use this version of the URL unconditionally.

That way we could also test that `expect( pathWithoutAMP ).not.toHaveAMPAutoAdsTag()` when AMP is active and configured to assert that the non-AMP version/URL of the page doesn't render it.

File: tests/e2e/plugins/apply-content-filters.php [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * @package   Google\Site_Kit
+ * @copyright 2020 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+
+/**
+ * Applying the filters outside of the loop causes the <amp-auto-ads> not be rendered on the single template
```
Comment (aaemnnosttv) [inline tests/e2e/plugins/apply-content-filters.php:16]:
```suggestion
 * Applying the `the_content` filter outside of the loop used to cause the <amp-auto-ads> not be rendered on the single template.
```

File: tests/phpunit/integration/Modules/AdSenseTest.php (hunk 8 above)
Comment (aaemnnosttv) [reply tests/phpunit/integration/Modules/AdSenseTest.php:212]: This test is yet to be added it seems.

File: tests/e2e/specs/modules/adsense/amp-auto-ads-tag.test.js [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+	it( 'is output in primary mode', async () => {
+		await setAMPMode( 'primary' );
+		await expect( '/hello-world' ).toHaveAMPAutoAdsTag();
+	} );
+
+	it( 'is output in secondary mode', async () => {
+		await setAMPMode( 'secondary' );
+		await expect( '/hello-world' ).toHaveAMPAutoAdsTag();
+	} );
```
Comment (aaemnnosttv) [inline tests/e2e/specs/modules/adsense/amp-auto-ads-tag.test.js:54]: As previously mentioned, let's also test here that the auto ads tag is _not_ output on the web version (not `?amp`) when in secondary AMP mode. 

For some reason it won't let me suggest on this file, but how about this for clarity?
* `is output in secondary mode in AMP context` _(existing test)_
* `is not output in secondary mode in web context` _(test to add)_

File: tests/e2e/specs/modules/adsense/amp-auto-ads-tag.test.js [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+ * Internal dependencies
+ */
+import {
+	setAMPMode,
+	setupAdSense,
+} from '../../../utils';
+
+describe( 'AMP <amp-auto-ads> tag', () => {
+	beforeAll( async () => {
+		await setupAdSense();
```
Comment (aaemnnosttv) [inline tests/e2e/specs/modules/adsense/amp-auto-ads-tag.test.js:37]: Looks great! Sooo much cleaner! 😄

**Review by aaemnnosttv — COMMENTED** [review]
Looking really good, just a few small things left to address and then this should be all set!

File: tests/e2e/matchers/to-have-amp-auto-ads-tag.js (hunk 11 above)
Comment (ryanwelcher) [reply tests/e2e/matchers/to-have-amp-auto-ads-tag.js:36]: @aaemnnosttv I originally had this in place however, the `?amp` was getting escaped so i went this route. I'll add a second optional boolean param that will add `amp` if `true`

File: tests/phpunit/integration/Modules/AdSenseTest.php [hunk 15]
```diff
@@ -225,6 +229,45 @@ public function block_on_consent_provider() {
...
+
+		// Confirm that the tag is not added if we're not in the loop.
+		$output = apply_filters( 'the_content', 'test content' );
+		$this->assertNotContains( 'data-ad-client="ca-pub-12345678"', $output );
+
+		// We need to fake the global to allow the hook to add the tag.
+		global $wp_query;
+		$wp_query->in_the_loop = true;
+
+		// Confirm that the tag is added if we're not in the loop.
```
Comment (aaemnnosttv) [inline tests/phpunit/integration/Modules/AdSenseTest.php:255]: This comment is a bit off, as it says we're still not in the loop.
```suggestion
		// Confirm that the tag is added when in the loop.
```

File: tests/phpunit/integration/Modules/AdSenseTest.php [hunk 16]
```diff
@@ -225,6 +229,45 @@ public function block_on_consent_provider() {
...
+		$output = apply_filters( 'the_content', 'test content' );
+		$this->assertContains( 'data-ad-client="ca-pub-12345678"', $output );
+	}
+
+	public function amp_auto_ads_tag_in_the_loop_provider() {
```
Comment (aaemnnosttv) [inline tests/phpunit/integration/Modules/AdSenseTest.php:260]: Super nitpick, but we recently opted to use the same naming as WP core for data providers which uses (albeit inconsistently) a `data_` prefix rather than a `_provider` suffix.
```suggestion
	public function data_amp_auto_ads_tag_in_the_loop() {
```

File: tests/phpunit/integration/Modules/AdSenseTest.php [hunk 17]
```diff
@@ -225,6 +229,45 @@ public function block_on_consent_provider() {
 		);
 	}
 
+	/**
+	 * @dataProvider amp_auto_ads_tag_in_the_loop_provider
```
Comment (aaemnnosttv) [inline tests/phpunit/integration/Modules/AdSenseTest.php:233]:
```suggestion
	 * @dataProvider data_amp_auto_ads_tag_in_the_loop
```

File: tests/e2e/specs/modules/adsense/amp-auto-ads-tag.test.js [hunk 18]
```diff
@@ -45,11 +46,16 @@ describe( 'AMP <amp-auto-ads> tag', () => {
...
 	it( 'is output in primary mode', async () => {
 		await setAMPMode( 'primary' );
-		await expect( '/hello-world' ).toHaveAMPAutoAdsTag();
+		await expect( createURL( '/hello-world' ) ).toHaveAMPAutoAdsTag();
 	} );
 
 	it( 'is output in secondary mode', async () => {
 		await setAMPMode( 'secondary' );
-		await expect( '/hello-world' ).toHaveAMPAutoAdsTag();
+		await expect( createURL( '/hello-world', 'amp' ) ).toHaveAMPAutoAdsTag();
```
Comment (aaemnnosttv) [inline tests/e2e/specs/modules/adsense/amp-auto-ads-tag.test.js:54]: Originally I was thinking the query could be passed through in the first argument to `createURL` but it gets URL encoded, so this is fine since there isn't a great way to pass the query separately. It's possible to pass through as an argument to the matcher itself, but `toHaveAMPAutoAdsTag( 'amp' )` would be weird.

**Review by aaemnnosttv — APPROVED** [review]
Super!
